collector/pipelinerun.go

```diff
@@ -0,0 +1,58 @@
+package collector
+
+import (
+	"context"
+	"time"
+
+	"github.com/go-kit/log"
+	"github.com/tektoncd/pipeline/pkg/apis/pipeline/v1beta1"
+
+	"k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/client-go/tools/record"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+)
+
+const (
+	contextTimeout = 300 * time.Second
+)
+
+func SetupPipelineRunReconcilerWithManager(mgr ctrl.Manager, logger log.Logger) error {
+	//TODO if/when we move off of the stand alone prometheus integration, we wil create metrics using the
+	// caching client from controller runtime
+	reconciler := &ReconcilePipelineRun{
+		client:        mgr.GetClient(),
+		scheme:        mgr.GetScheme(),
+		eventRecorder: mgr.GetEventRecorderFor("PipelineRun"),
+		logger:        logger,
+	}
+	return ctrl.NewControllerManagedBy(mgr).For(&v1beta1.PipelineRun{}).Complete(reconciler)
+}
+
+type ReconcilePipelineRun struct {
+	client        client.Client
+	scheme        *runtime.Scheme
+	eventRecorder record.EventRecorder
+	logger        log.Logger
+}
+
+func (r *ReconcilePipelineRun) Reconcile(ctx context.Context, request reconcile.Request) (reconcile.Result, error) {
```

Comment:
Why do we need an empty controller? Is there an easier way to get a client backed by a lister cache?

Reply:
Because I have another story after this one that is going to add the throttled by quota and throttled on  node resources constraints that I've demo'ed in the past.  That checks conditions from the Pod that tekton transfers to the PipelineRun.

Hence I'm not interested in sorting out the client backed by a PipelineRun lister cache only thing that will only be in place for a short while.  Having this here could also help with debug in between the 2 stories if the need arises.

I'll add a TODO comment explaining this will be filled in soon.

Reply:
Forgot - I can do the new/upcoming stats with the prometheus/collect flow (vs. counter.Inc() etc.).

And if the godoc is accurate, `ctrl.NewControllerManagedBy(mgr).For(&v1beta1.PipelineRun{})` may be sufficient for creating the watch/lister cache, where bypassing the call to `Complete(reconciler)` means we will not get called on events.  I'll see if I can cobble together an experiment tomorrow to prove that.

Reply:
Had some fun with the debugger today @adambkaplan 

Using the code as is, I set break points in 
- k8s.io/client-go/tools/cache/shared_informer.go:NewSharedIndexInformer
- k8s.io/client-go/tools/cache/reflector.go:ListAndWatch

And got hits for both when bringing up the exporter.

Then, when I changed 

`ctrl.NewControllerManagedBy(mgr).For(&v1beta1.PipelineRun{}) .Complete(reconciler)` 

to 

`ctrl.NewControllerManagedBy(mgr).For(&v1beta1.PipelineRun{})`

I got no hit.  But when I followed that with 

```
ctrl.NewControllerManagedBy(mgr).For(&v1beta1.PipelineRun{})
prList := v1beta1.PipelineRunList{}
opts := &client.ListOptions{Namespace: metav1.NamespaceAll}
mgr.GetClient().List(context.Background(), &prList, opts)
```

and restarted, I got hits on those break points again.

In hindsight, I now realize this lines up with controller-runtime's non-caching client options (`options.ClientDisableCacheFor`) where if you don't set that, controller-runtime's use of dynamic client includes setting up a caching client by default.

Bottom line,  I'll lose the empty reconciler :-)

Reply:
actually I do not need `ctrl.NewControllerManagedBy(mgr).For(&v1beta1.PipelineRun{})` as well